### PR TITLE
Multi Thread coverage merging to use full machine

### PR DESCRIPTION
**Pull request by dothebart**
- add psutil to containers used for coverage runs
- remove serial fish logic to join coverage directories
- create python script to do that:
  - gather all existing coverage directories
  - build a matrix of a + b => c; rm a, b until only one coverage set remains
  - spawn queue-fed worker threads, that will run combine jobs. 
  - for each layer of reduction wait for jobs to finish so we don't overlap. 

It seems this gives a ~7 times speedup, bringing us down from ~105 minutes to ~15.

### CODE REVIEW

File: scripts/coverage.fish [hunk 1]
```diff
@@ -2,34 +2,9 @@
...
-
-and if test $c -eq 1
-  mv combined/1 combined/result
-else if test $c -eq 2
-  mv combined/2 combined/result
-end
-
-and rm -rf combined/1 combined/2 /tmp/gcno
+and rm -rf /work/combined
+and mkdir /work/combined
```
Comment (KVS85) [inline scripts/coverage.fish:6]:
```suggestion
and rm -rf ./combined; and mkdir ./combined
```

File: scripts/coverage.fish [hunk 2]
```diff
@@ -2,34 +2,9 @@
...
+and python3 "$WORKSPACE/jenkins/helper/aggregate_coverage.py" /work/gcov /work/combined/result
```
Comment (KVS85) [inline scripts/coverage.fish:7]: Check whether WORKSPACE is empty:
- yes: use `$WORKDIR/jenkins/helper/aggregate_coverage.py`
- no: use `$WORKSPACE/jenkins/helper/aggregate_coverage.py`

Use `$INNERWORKDIR/gcov $INNERWORKDIR/combined/result`.

**Review by KVS85 — CHANGES_REQUESTED** [review]

**Comment by KVS85** [comment]
- [ ] https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/28110/
- [ ] https://jenkins01.arangodb.biz/job/arangodb-ANY-coverage/46/

**Review by KVS85 — APPROVED** [review]

**Comment by KVS85** [comment]
- [ ] https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/28129/
- [ ] https://jenkins01.arangodb.biz/job/arangodb-ANY-coverage/48/

File: jenkins/helper/aggregate_coverage.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+SUCCESS = True
+
+# pylint disable=global-variable-not-assigned
+
+class GcovMerger(ArangoCLIprogressiveTimeoutExecutor):
+    """configuration"""
+
+    def __init__(self, jobs, site_config):
+        self.identifier = jobs[0]
+        self.job = jobs
```
Comment (alexbakharew) [inline jenkins/helper/aggregate_coverage.py:29]: Please use `self.jobs` as we have array of jobs

File: jenkins/helper/aggregate_coverage.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        self.job_parameters = ['merge', jobs[0], jobs[1], '-o', jobs[2]]
+        self.params = None
+        super().__init__(site_config, None)
+
+    def launch(self):
+       # pylint: disable=R0913 disable=R0902 disable=broad-except
+        """ gcov merger """
+        print('------')
```
Comment (alexbakharew) [inline jenkins/helper/aggregate_coverage.py:37]: Do we need here this output here?

File: jenkins/helper/aggregate_coverage.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        verbose = True
+        self.params = make_default_params(verbose, 111)
+        print(self.params)
```
Comment (alexbakharew) [inline jenkins/helper/aggregate_coverage.py:40]: And here?

**Review by alexbakharew — APPROVED** [review]
LGTM. But we need to get rid of redundant output.

**Comment by KVS85** [comment]
- [ ] https://jenkins01.arangodb.biz/job/arangodb-ANY-coverage/49/

**Review by naushniki — APPROVED** [review]

**Comment by KVS85** [comment]
- [ ] https://jenkins01.arangodb.biz/job/arangodb-ANY-coverage/50/

File: jenkins/helper/aggregate_coverage.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        launch_worker(cfg)
+        worker_count -= 1
+    for one_job_set in jobs:
+        count = 0
+        for one_job in one_job_set:
+            JOB_QUEUE.put(one_job)
+            count += 1
+        print(f'waiting for jobset {count} to finish')
+        while count > 0:
+            print('.')
```
Comment (alexbakharew) [inline jenkins/helper/aggregate_coverage.py:155]: Is this output required?

**Review by alexbakharew — COMMENTED** [review]
Additional comments

File: jenkins/helper/aggregate_coverage.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+            this_subjob = [str(sub_jobs.pop()),
+                           str(sub_jobs.pop()),
+                           str(last_output)]
+            jobs[count].append(this_subjob)
+            next_jobs.append(this_subjob[2])
+            jobcount += 1
+        count += 1
+        if len(sub_jobs) > 0:
+            next_jobs.append(sub_jobs.pop())
+        sub_jobs = next_jobs
```
[the remark this answers is not in the document]
Comment (dothebart) [reply jenkins/helper/aggregate_coverage.py:140]: if you have an odd number of directories, one remains for the next layer of the tree.

**Comment by dothebart** [comment]
I will clean up the output and the variable names in the next iteration.

**Pull request merged**